proposals/span-safety.md

```diff
@@ -0,0 +1,238 @@
+
+The attribute will be added to common base libraries such as `mscorlib`. In a case if the attribute is not available, compiler will generate an internal one similarly to other embedded-on-demand attributes such as `IsReadOnlyAttribute`.
+
+An additional measure will be taken to prevent the use of ref-like structs in compilers not familiar with the safety rules (this includes C# compilers prior to the one in which this feature is implemented). 
+
+Having no other good alternatives that work in old compilers without servicing, an `Obsolete` attribute with a known string will be added to all ref-like structs. Compilers that know how to use ref-like types will ignore this particular form of `Obsolete`.
+
+A typical matadata representation: 
+
+```C#
+    [IsRefLike]
+    [Obsolete("Types with embedded references are not supported in this version of your compiler.")]
+    public struct TwoSpans<T>
+    {
+       . . . . 
+    }
+```
+
+NOTE: it is not the goal to make it so that any use of ref-like types on old compilers fails 100%. That is hard to achieve and is not strictly necessary. For example there would always be away to get around the `Obsolete` using dynamic code or, for example, creating an array of ref-like types through reflection.
+
+In particular, if user wants to actually put an `Obsolete` attribute on a ref-like type, we will have no choice other than not emitting the predefined one since `Obsolete` attribute cannot be applied more than once..  
+(TODO: We should consider giving a warning in such scenario, informing user of the danger.)
+
+## Stack-referring spans
+
+An additional desirable enhancement for `Span<T>` is to allow referring to data in the local stack frame - individual local variables or `stackalloc`-ed arrays.
+
+The tricky part here is that a scenario when an instance of a `Span<T>` outlives the referred data would lead to undefined behavior, including type-safety violations and heap corruptions. Therefore, in order to allow stack-referring spans in safe code, we need additional rules. 
+
+The design goal here is to make rules for stack-referring spans "pay-for-play".  
+**A person who does not intend to trade in stack-referring spans should be able to stop reading right here.**
+ 
+NOTE: nothing will prevent a possibility to use Spans that wrap pointers to stack-allocated arrays in _unsafe_ code as long as it is possible to wrap a pointer in a span. As per usual rules of _unsafe_, it would be responsibility of the user to not cause GC holes.
+
+## Stack-referring spans in safe code.
+
+There are two main scenarios where stack-referring spans could come into being.  
+- Stackalloc spans.
+  To enable safe wrapping of stack allocated memory in a span, we will need a special syntax that would not allow intermediate pointer to escape.
+  A preliminary solution is to use target-typing.
+
+```cs
+  //This usage of stackalloc does not require unsafe 
+  Span<int> sp = stackalloc int[100]; 
```

Comment:
This is interesting. It doesn't require unsafe but the resulting code is unverifiable. It would seem that we're moving towards a system where the managed compiler(s) are responsible for ensuring type safety and not the runtime.